device/transport/mqtt/src/mqtt.ts

```diff
@@ -100,43 +100,68 @@ export class Mqtt extends EventEmitter implements Client.Transport, StableConnec
     this.on('newListener', (eventName) => {
       // if the event is a 'method' event then eventName is in the format
       // 'method_{method name}'
-      const topic = eventName.indexOf('method_') === 0 ?
-                      this._topics.method :
-                      eventName === 'message' ?
-                      this._topics.message :
-                        null;
-      if (!!topic) {
-        // increment listeners count for this topic
+      if (eventName.indexOf('method_') === 0) {
+        const topic = this._topics.method;
         ++(topic.listenersCount);
-
-        // lazy-init MQTT subscription
         if (topic.subscribed === false && topic.subscribeInProgress === false) {
           // Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_003: [ When a listener is added for the message event, the topic should be subscribed to. ]
           // Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_13_002: [ When a listener is added for the method event, the topic should be subscribed to. ]
-          this._setupSubscription(topic);
+          this.enableMethods((err) => {
+            if (err) {
+              debug('error setting up subscription for ' + topic.name + ': ' + err.toString());
+            }
+          });
+        } else {
+          debug('subscription for methods is already set up');
         }
+      } else if (eventName === 'message') {
+        const topic = this._topics.message;
+        ++(topic.listenersCount);
+        if (topic.subscribed === false && topic.subscribeInProgress === false) {
+          // Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_16_003: [ When a listener is added for the message event, the topic should be subscribed to. ]
+          // Codes_SRS_NODE_DEVICE_MQTT_RECEIVER_13_002: [ When a listener is added for the method event, the topic should be subscribed to. ]
+          this.enableC2D((err) => {
+            if (err) {
+              debug('error setting up subscription for ' + topic.name + ': ' + err.toString());
+            }
+          });
+        } else {
+          debug('subscription for C2D messages is already set up');
+        }
+      } else {
+        debug('new listener for which there\'s nothing to subscribe to: ' + eventName);
       }
     });
 
     this.on('removeListener', (eventName) => {
       // if the event is a 'method' event then eventName is in the format
       // 'method_{method name}'
-      const topic = eventName.indexOf('method_') === 0 ?
-                  this._topics.method :
-                  eventName === 'message' ?
-                    this._topics.message :
-                    null;
-      if (!!topic) {
-        debug('removing listener for topic: ' + JSON.stringify(topic));
-        // decrement listeners count for this topic
+      if (eventName.indexOf('method_') === 0) {
+        const topic = this._topics.method;
+        --(topic.listenersCount);
+        if (topic.listenersCount === 0 && topic.subscribed === true) {
```

Comment:
>topic.subscribed === true [](start = 42, length = 25)

edge case: what if topic.subscribeInProgress is true ?

Reply:
should be fine. we don't really care about sending multiple subscribe packets. if anything, I'd be worried that the first failed and we don't retry because it's already in progress.

we need better state management all around here though.. totally agree with the idea... out of scope for this PR though.

---
In reply to: [143822908](https://github.com/Azure/azure-iot-sdk-node/pull/103#discussion_r143822908) [](ancestors = 143822908)